### ISSUE TITLE
Using Math.random breaks context.df.currentUtcDateTime in Retry context

### ISSUE THREAD

**Issue by kepikoi** [Azure/azure-functions-durable-js#224]
Delaying a durable orchestrator using for variable amount of seconds via `Math.random` causes the loop to stop working after the first retry:
 
## defer random amount of seconds using Math.random

```ts
export const defer = orchestrator(function* (context) {
    const current = context.df.currentUtcDateTime;
    console.log("current", current);
    const pauseMs = (5 + Math.random()) * 1e3; // min. 5 max. 15s
    const nextDate = new Date(current.getTime() + pauseMs);
    console.log("next", nextDate);
    yield context.df.createTimer(nextDate); // sleep and retry

    return context.df.continueAsNew(undefined);
});
```

```
[2020-11-16T15:08:11.046] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=bb5ebb4f-8ee5-47b9-a896-279cb6c0a320)
[2020-11-16T15:08:11.074] current 2020-11-16T15:08:11.040Z
[2020-11-16T15:08:11.085] next 2020-11-16T15:08:17.004Z
[2020-11-16T15:08:11.106] Executed 'Functions.DO_aml-defer' (Succeeded, Id=bb5ebb4f-8ee5-47b9-a896-279cb6c0a320, Duration=62ms)
[2020-11-16T15:08:19.493] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=270f84af-820f-4f03-95fb-4ba30b0dd886)
[2020-11-16T15:08:19.512] current 2020-11-16T15:08:11.040Z
[2020-11-16T15:08:19.516] next 2020-11-16T15:08:16.160Z
[2020-11-16T15:08:19.529] Executed 'Functions.DO_aml-defer' (Succeeded, Id=270f84af-820f-4f03-95fb-4ba30b0dd886, Duration=54ms)

does nothing after this
```

Basically the same code but without `Math.radom` works as expected:

## defer set amount of seconds

```ts
export const defer = orchestrator(function* (context) {
    const current = context.df.currentUtcDateTime;
    console.log("current", current);
    const pauseMs = 10 * 13; // 10s
    const nextDate = new Date(current.getTime() + pauseMs);
    console.log("next", nextDate);
    yield context.df.createTimer(nextDate); // sleep and retry

    return context.df.continueAsNew(undefined);
});
```

```
[2020-11-16T15:20:25.149] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=7bc703ca-0d89-4c2d-a061-796bc1e2878d)
[2020-11-16T15:20:25.164] current 2020-11-16T15:20:25.144Z
[2020-11-16T15:20:25.167] next 2020-11-16T15:20:35.144Z
[2020-11-16T15:20:25.197] Executed 'Functions.DO_aml-defer' (Succeeded, Id=7bc703ca-0d89-4c2d-a061-796bc1e2878d, Duration=50ms)
[2020-11-16T15:20:39.223] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=737b0d49-f2dd-4aa7-8645-79624e370c86)
[2020-11-16T15:20:39.237] current 2020-11-16T15:20:25.144Z
[2020-11-16T15:20:39.242] next 2020-11-16T15:20:35.144Z
[2020-11-16T15:20:39.259] Executed 'Functions.DO_aml-defer' (Succeeded, Id=737b0d49-f2dd-4aa7-8645-79624e370c86, Duration=53ms)
[2020-11-16T15:20:39.284] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=6d4456a4-e49e-42ee-9607-5ed9e798da2e)
[2020-11-16T15:20:39.296] current 2020-11-16T15:20:39.281Z
[2020-11-16T15:20:39.299] Executed 'Functions.DO_aml-defer' (Succeeded, Id=6d4456a4-e49e-42ee-9607-5ed9e798da2e, Duration=16ms)
[2020-11-16T15:20:39.303] next 2020-11-16T15:20:49.281Z
[2020-11-16T15:20:52.878] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=cefcc82c-fabe-4ce1-af40-cc0b059184de)
[2020-11-16T15:20:52.902] current 2020-11-16T15:20:39.281Z
[2020-11-16T15:20:52.907] next 2020-11-16T15:20:49.281Z
[2020-11-16T15:20:52.910] Executed 'Functions.DO_aml-defer' (Succeeded, Id=cefcc82c-fabe-4ce1-af40-cc0b059184de, Duration=66ms)
[2020-11-16T15:20:52.925] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=bc93a8c2-2f4a-4355-a8bc-29d2ac7931cd)
[2020-11-16T15:20:52.944] current 2020-11-16T15:20:52.921Z
[2020-11-16T15:20:52.950] Executed 'Functions.DO_aml-defer' (Succeeded, Id=bc93a8c2-2f4a-4355-a8bc-29d2ac7931cd, Duration=25ms)
[2020-11-16T15:20:52.950] next 2020-11-16T15:21:02.921Z
[2020-11-16T15:21:05.808] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=1c18fd22-a3bd-4db0-bb62-8026a72cea66)
[2020-11-16T15:21:05.815] current 2020-11-16T15:20:52.921Z
[2020-11-16T15:21:05.817] next 2020-11-16T15:21:02.921Z
[2020-11-16T15:21:05.817] Executed 'Functions.DO_aml-defer' (Succeeded, Id=1c18fd22-a3bd-4db0-bb62-8026a72cea66, Duration=10ms)
[2020-11-16T15:21:05.825] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=cf9a2e41-47fc-46b8-a494-aa1a6e00b2b8)
[2020-11-16T15:21:05.835] current 2020-11-16T15:21:05.824Z
[2020-11-16T15:21:05.836] Executed 'Functions.DO_aml-defer' (Succeeded, Id=cf9a2e41-47fc-46b8-a494-aa1a6e00b2b8, Duration=11ms)
[2020-11-16T15:21:05.837] next 2020-11-16T15:21:15.824Z
[2020-11-16T15:21:19.226] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=49f700ae-12a4-476e-938d-3e632061195f)
[2020-11-16T15:21:19.236] current 2020-11-16T15:21:05.824Z
[2020-11-16T15:21:19.240] next 2020-11-16T15:21:15.824Z
[2020-11-16T15:21:19.240] Executed 'Functions.DO_aml-defer' (Succeeded, Id=49f700ae-12a4-476e-938d-3e632061195f, Duration=15ms)
[2020-11-16T15:21:19.249] Executing 'Functions.DO_aml-defer' (Reason='(null)', Id=b35bf156-7e5e-40c8-9edf-7e9f8d3086e5)
[2020-11-16T15:21:19.260] current 2020-11-16T15:21:19.247Z
[2020-11-16T15:21:19.262] Executed 'Functions.DO_aml-defer' (Succeeded, Id=b35bf156-7e5e-40c8-9edf-7e9f8d3086e5, Duration=13ms)
[2020-11-16T15:21:19.262] next 2020-11-16T15:21:29.247Z

 ... runs forever
```

I'm running

"durable-functions": "^1.4.3",

node -v 
v12.18.3

func -v 
3.0.2931

[System.Environment]::OSVersion.Version

Major  Minor  Build  Revision
-----  -----  -----  --------
10     0      18363  0

**Comment by davidmrdavid** [comment]
Hi @kepikoi! 

Thanks for reaching out. I think I know what's going on. In general, Durable Functions imposes some coding constraints on orchestrators due to its replayability semantics. In this case, using `math.random` is violating the determinism contraint, which requires orchestrators to use deterministic APIs. To call a non-deterministic API, you need to do it inside an activity. 

You can read more about this here: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-code-constraints#using-deterministic-apis

In short, the fix here should be to move your random function inside an activity function that returns its value. Let me know if that doesn't make sense!

Also, just a friendly reminder that activities need to be idempotent (safe to do more than once), which is another coding constraint.

:zap: :zap:

**Comment by kepikoi** [comment]
Ok, that was a silly question - it's right there in the docs 😅 Thanks a lot!

**Comment by davidmrdavid** [comment]
No silly questions, thank you for reaching out :)
